### ISSUE TITLE
[Symfony] Why is there no amLoggedAs?

### ISSUE THREAD

**Issue by krukru** [Codeception/Codeception#5699]
#### What are you trying to achieve?

I am trying to write a functional test for a symfony app. The app has a traditional login form based authentication system. 
I would like to avoid logging in via submitForm as this slows my tests down and I need to know passwords for the test users I am logging in with (this issue is minor in comparison to test execution speed).

Yii2 and Laravel Modules have `amLoggedInAs` and `amLoggedIn` methods which accomplish exactly what I need, but Symfony has none that I can find useful.

Is there some framework limitation why this does not exist? Would it be possible to create this method using https://symfony.com/doc/current/security/guard_authentication.html#manually-authenticating-a-user ?

**Comment by Naktibalda** [comment]
> Why is there no amLoggedAs?

Because noone implemented it.

> Is there some framework limitation why this does not exist?

This is framework specific question and we have no maitainer for Symfony module currently.

**Comment by monotonesol** [comment]
You could write it by yourself, it should be pretty similar to: 
`<?php

namespace TestSupport\Helper;

use Codeception\Module;
use Codeception\Module\Symfony;
use Core\UserBundle\Entity\User;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
use Symfony\Component\Security\Http\Event\InteractiveLoginEvent;
use Symfony\Component\Security\Http\SecurityEvents;

/**
 * Class Login
 * @package TestSupport\Helper
 */
class Login extends Module
{
    /**
     * @var Symfony
     */
    private $symfony;

    public function _initialize($settings = [])
    {
        $this->symfony = $this->getModule('Symfony');
    }

    /**
     * @param \Core\UserBundle\Entity\User $user
     */
    public function loggedInAs(User $user)
    {
        $session = $this->symfony->grabService('session');

        $firewall = 'main';

        $token = new UsernamePasswordToken($user, null, $firewall, array_merge($user->getRoles(), ['ROLE_NULL']));
        $session->set('_security_'.$firewall, serialize($token));

        $this->symfony->setCookie($session->getName(), $session->getId());
        $request = Request::create('/login');
        $request->setSession($session);
        $loginEvent = new InteractiveLoginEvent($request, $token);
        $this->symfony->grabService('security.token_storage')->setToken($token);
        $this->symfony->grabService('event_dispatcher')->dispatch(SecurityEvents::INTERACTIVE_LOGIN, $loginEvent);
        $session->save();
    }
}
`

**Comment by c33s** [comment]
i use the following based upon the [symfony doc for testing][1]  
`tests/_support/Helper/Functional.php`:
```php
<?php
namespace Tests\Helper;

use Tests\Traits\CodeceptionSymfonyTrait;
use Symfony\Component\HttpFoundation\Session\Session;
use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
use Symfony\Component\BrowserKit\Cookie;

class Functional extends \Codeception\Module
{
    use CodeceptionSymfonyTrait;
    /**
     * Auto Login a User without creating it in the Database with writing directly to session.
     *
     * doesn't work with Groups
     *
     * @param string $user
     * @param string $firewallContext
     * @param array  $roles
     */
    public function loggedInAs($user = 'admin', $firewallContext = 'main', $roles = ['ROLE_SUPER_ADMIN', 'ROLE_ADMIN', 'ROLE_USER', 'ROLE_SONATA_ADMIN'])
    {
        /** @var Session $session */
        $session = $this->getContainer()->get('session');
        $client = $this->getClient();

        $token = new UsernamePasswordToken($user, null, $firewallContext, $roles);
        $session->set('_security_'.$firewallContext, serialize($token));
        $session->save();
        $cookie = new Cookie($session->getName(), $session->getId());
        $client->getCookieJar()->set($cookie);
    }

    public function loggedInAsAdmin()
    {
        $this->loggedInAs('admin', 'main', ['ROLE_ADMIN']);
    }
    public function loggedInAsEditor()
    {
        $this->loggedInAs('editor', 'main', ['ROLE_EDITOR']);
    }
```

[1]: https://symfony.com/doc/current/testing/http_authentication.html#creating-the-authentication-token

**Comment by TavoNiievez** [comment]
Hey guys, I just submitted a PR adding this feature. Thanks for sharing your implementations.

**Comment by Naktibalda** [comment]
amLoggedInAs method was implemented in module-symfony 1.2.0
